### PR TITLE
Implement Docker Base Image Mapper

### PR DESCRIPTION
**Pull request by c-mueller**
For some information about what the mapper does please consult #218 and [BaseImageMapper Spec](https://github.com/StuPro-TOSCAna/TOSCAna/blob/2d72d374fff07c6ff694930e828df4f534f47b5d/docs/dev/plugins/kubernetes/dockerfile-mapping/base-image-mapper.md)

The following tasks will have to be done to complete this PR:
- [x] Create Special configuration to allow fast loading in tests. (Probably by injecting a map of data before the context gets created)
- [x] Replace `OSCapability` with the one from the model, or at least convert the one from the model to `OSCapability`
- [x] Document the Code
- [x] Update Specification document

### CODE REVIEW

File: server/pom.xml [hunk 1]
```diff
@@ -24,17 +24,24 @@
     </repositories>
 
     <dependencies>
-        <dependency>
```
Comment (nfode) [inline server/pom.xml]: Why did you remove this dependency?

**Comment by c-mueller** [comment]
The code is ready to review, Javadoc will be added soon. The merge cannot take place until the Model has been merged into master

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            architecture = ARCHITECTURE_MAP.get(capability.getArchitecture().get());
+        } else {
+            //Default to AMD64 (aka. x86_64)
+            architecture = "amd64";
+        }
+        return architecture;
+    }
+
+    private boolean isDistributionNameValid(OSCapability capability) {
+        if (hasDistribution(capability)) {
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java]: `if (capability.getDistribution().isPresent(){
} else etc..`

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    private Map<String, DockerImage> imageMap = new HashMap<>();
+
+    /**
+     A array conatinig the base images that this Base Image Mapper tries to map to
+     this usually is <code>DockerBaseImages.values()</code>
+     */
+    private DockerBaseImages[] baseImages;
+
+    @Autowired
+    public BaseImageMapper(DockerBaseImages[] dockerBaseImages) {
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java]: This class is huge. You should split it up.
I propose you extract the logic regarding the versioning into its own class.
Furthermore, I propose to use an external library for semantic versioning parsing - it's an often solved problem and implementing it by ourselves means more brittle code to maintain.

File: server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/util/MapperTestConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ Import this annotation if you need the base image mapper in your integration tests
+ it will load a file from the test resources in order to perform the mappings.
+ To keep loading times at a minimum
+ add it using the following annotation on the class
+ <code>@ContextConfiguration(classes = {MapperTestConfiguration.class})</code>
+ and
+ <code>@ActiveProfiles({"bim"})</code>
+ */
+@Profile("bim")
```
Comment (hnicke) [inline server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/util/MapperTestConfiguration.java]: bim? What does this stand for?

File: server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    @Test
+    public void testConversion() throws Exception {
+        OsCapability capability = builder()
+            .architecture(Architecture.x86_64)
+            .distribution(Distribution.UBUNTU)
+            .type(Type.LINUX)
+            .version("16.04").build();
+
+        OSCapability c = new OSCapability(capability);
+        assertEquals("x86_64", c.getArchitecture().get());
```
Comment (hnicke) [inline server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtilsTest.java]: Maybe get rid of the hardcoded strings.

File: server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTest.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        IOUtils.copy(in, out);
+        BaseImageMapper baseImageMapper = new BaseImageMapper(new DockerBaseImages[] {
+            ALPINE,
+            DEBIAN,
+            UBUNTU
+        });
+        DataContainer data = mapper.readValue(new String(out.toByteArray()), DataContainer.class);
+
+        ReflectionUtils.setPrivateField(baseImageMapper, "imageMap", data.data);
```
Comment (hnicke) [inline server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTest.java]: Testing private fields is an anti pattern. Tests should only test the external behaviour. If internal implementation changes, tests break. I suggest you get rid of the Reflection Util class and change imageMap visibility to default (package visiblity).

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java (hunk 2 above)
Comment (c-mueller) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java]: just look what hasDistribution does :)

File: server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/util/MapperTestConfiguration.java (hunk 4 above)
Comment (c-mueller) [reply server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/util/MapperTestConfiguration.java]: baseimagemapper

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/OSCapability.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import org.opentosca.toscana.model.capability.OsCapability;
+
+/**
+ This class represents a TOSCA OS Capability
+ Because the Model is not yet available this class is used to model the
+ OS Capability
+ <p>
+ TODO: Replace with real OsCapability class once available
+ */
+public class OSCapability {
```
[the remark this answers is not in the document]
Comment (c-mueller) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/OSCapability.java]: this is one of the next steps an the main reason why this PR is still WIP

File: server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTest.java (hunk 6 above)
Comment (c-mueller) [reply server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperTest.java]: This is not testing, this is used to bypass the Download mechanism.

File: server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/util/MapperTestConfiguration.java (hunk 4 above)
Comment (hnicke) [reply server/src/test/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/util/MapperTestConfiguration.java]: Maybe write `base-image-mapper` then. I guess in 4 weeks everyone has forgotten what `bim` is.

File: docs/dev/plugins/kubernetes/dockerfile-mapping/base-image-mapper.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+When the application launches the BaseImageMapper trys to retrieve a list of all tags for the supported docker base images.
+The BaseImageMapper will support the following base images in the first revision:
+- `ubuntu`
+- `debian`
+- `alpine`
+- `centos`
+- `opensuse`
+- `busybox`
+
+On launch the application will download the latest tags from dockerhub. Once this process is done the Downloaded tag list will be stored in the datadirectory provided by the server. This serialisation will be done as JSON by the use of the Jackson (`ObjectMapper`) library. Once the download is done once A update of the Mapping files is only done every 24 hours (will be a changable property). The Key to change this value is `toscana.docker.base-image-mapper.update-interval`
```
Comment (hnicke) [inline docs/dev/plugins/kubernetes/dockerfile-mapping/base-image-mapper.md]: fix sentence: `Once the download is done once A update ..`

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Profile;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+/**
+ This class allows the mapping of OSCapabilities (form a tosca model) to a docker base image.
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java]: `OSCapabilities` -> `OsCapabilities`, `form` -> `from`

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/DockerBaseImages.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+public enum DockerBaseImages {
+    UBUNTU("library", "ubuntu", DOCKER_HUB_URL),
+    DEBIAN("library", "debian", DOCKER_HUB_URL),
+    CENTOS("library", "centos", DOCKER_HUB_URL),
+    ALPINE("library", "alpine", DOCKER_HUB_URL),
+    BUSYBOX("library", "busybox", DOCKER_HUB_URL),
+    OPEN_SUSE("library", "opensuse", DOCKER_HUB_URL);
+
+    private String username;
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/DockerBaseImages.java]: make these dudes final please

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperConstants.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+import static org.opentosca.toscana.model.capability.OsCapability.Architecture.x86_32;
+import static org.opentosca.toscana.model.capability.OsCapability.Architecture.x86_64;
+
+public class MapperConstants {
+    public static final String DEFAULT_IMAGE_PATH = "library/ubuntu:latest";
+    public static final String DEFAULT_IMAGE_DISTRO = "ubuntu";
+    public static final String DOCKER_HUB_URL = "https://registry.hub.docker.com";
+    public static final Map<Architecture, String> ARCHITECTURE_MAP;
+    public static final String VERSION_DELIMITER_REGEX = "\\.";
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperConstants.java]: Guess this one is deprecated

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+import static org.opentosca.toscana.plugins.kubernetes.docker.mapper.MapperConstants.DEFAULT_IMAGE_PATH;
+import static org.opentosca.toscana.plugins.kubernetes.docker.mapper.MapperUtils.TAG_COMPARATOR_MINOR_VERSION;
+import static org.opentosca.toscana.plugins.kubernetes.docker.mapper.MapperUtils.anythingSet;
+
+/**
+ This class wraps the mapping functionality of the Base image mapper
+ */
+class MapperEngine {
+
+    private Map<String, DockerImage> imageMap;
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java]: final :)

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+                int j = i;
+                DockerImageTag prev = null;
+                DockerImageTag t = null;
+                Semver tv;
+                do {
+                    prev = t;
+                    t = possibleTags.get(j);
+                    tv = t.toVersion();
+                    j++;
+                } while (tv.getMinor() == version.getMinor() && j < possibleTags.size());
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java]: `Integer` (in contrast to `int`) cannot be compared with `==`, they are complex objects, therefore you need to use `equals`.

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+            return img.getUsername() + "/" + img.getRepository() + ":" + imgTag.getName();
+        } else {
+            throw new UnsupportedOperationException("Could not map to base image. " +
+                "The image '" + img.getUsername() + "/" + img.getRepository() + ":" + tag + "' " +
+                "does not support the architecture " + architecture);
+        }
+    }
+
+    /**
+     Converts the architecture of the capability of a architecture supported by docker
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java]: ...to an architecture.. ?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+     */
+    private Optional<String> getCapabilityArchitecture(OsCapability capability) {
+        return Optional.ofNullable(ARCHITECTURE_MAP.get(capability.getArchitecture().orElse(x86_64)));
+    }
+
+    /**
+     Checks if the distribution name is valid (known)
+     */
+    private boolean isDistributionNameValid(OsCapability capability) {
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java]: maybe rename to `hasKnownDistributionName` && delete javadoc comment

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        if (capability.getDistribution().isPresent()) {
+            String distroName = convertDistributionName(capability);
+            DockerImage image = getDockerImage(distroName);
+            return image != null;
+        }
+        return false;
+    }
+
+    private DockerImage getDockerImage(String distroName) {
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperEngine.java]: This method is unnecessary

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+        Comparator.comparing(DockerImageTag::toVersion);
+
+    public static boolean isAnyOptionalPresent(Optional... optionals) {
+        for (Optional optional : optionals) {
+            if (optional.isPresent()) {
+                return true;
+            }
+        }
+        return false;
+    }
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java]: Could be replaced with `        Arrays.stream(optionals).anyMatch(optional -> optional.isPresent());`

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+    public static boolean anythingSet(OsCapability capability) {
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java]: This method could alternatively be placed in OsCapability itself.

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/api/DockerRegistry.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+ <p>
+ To connect to the registry retrofit is used and wrapped in this class.
+ */
+public class DockerRegistry {
+
+    private static final Logger logger = LoggerFactory.getLogger(DockerRegistry.class);
+
+    private RegistryService service;
+
+    private String baseUrl;
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/api/DockerRegistry.java]: this dude wants to be final when he's grown up

File: server/src/test/java/org/opentosca/toscana/core/BaseTest.java [hunk 20]
```diff
@@ -6,7 +6,9 @@
 import org.junit.Rule;
 import org.mockito.junit.MockitoJUnit;
 import org.mockito.junit.MockitoRule;
+import org.springframework.test.context.ActiveProfiles;
 
+@ActiveProfiles({"test"})
```
Comment (hnicke) [inline server/src/test/java/org/opentosca/toscana/core/BaseTest.java]: This profile name is quite generic, could you find a better one?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java (hunk 18 above)
Comment (c-mueller) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java]: I think its better to keep it here because no one other then the Mapper will probably use this.

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+    public static final Comparator<DockerImageTag> TAG_COMPARATOR_MINOR_VERSION =
+        Comparator.comparing(DockerImageTag::toVersion);
+
+    public static boolean isAnyOptionalPresent(Optional... optionals) {
+        return Arrays.stream(optionals).anyMatch(Optional::isPresent);
+    }
+
+    public static boolean anythingSet(OsCapability capability) {
+        return isAnyOptionalPresent(
```
Comment (hnicke) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/MapperUtils.java]: It's the only use of `isAnyOptionalPresent`, so I suggest to merge both methods.

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+@Profile("!" + Profiles.EXCLUDE_BASE_IMAGE_MAPPER)
+@SuppressWarnings("ConstantConditions")
+public class BaseImageMapper {
+
+    private static final Logger logger = LoggerFactory.getLogger(BaseImageMapper.class);
+
+    /**
+     Stores the update intervall for the base Image mappings (in hours)
+     Taken from the property value:
+     <c
```
Comment (nfode) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java]: ?

File: server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+                addImagesForType(baseImage, tags);
+            }
+        }
+        lastUpdate = System.currentTimeMillis();
+        logger.info("Mappings have been updated. The next update will be executed in approx. {} hours", updateInterval);
+    }
+
+    /**
+     This is the spring scheduled job used to perform the updates after the frist one
+     <p>
```
Comment (nfode) [inline server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java]: please close the paragraph

Comment (c-mueller) [reply server/src/main/java/org/opentosca/toscana/plugins/kubernetes/docker/mapper/BaseImageMapper.java]: if i recall correctly thats not needed when messing with javadocs

File: docs/dev/plugins/kubernetes/dockerfile-mapping/base-image-mapper.md [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+- `busybox`
+
+On launch the application will download the latest tags from DockerHub.
+Once this process is done the Downloaded tag list will be stored in memory.
+update of the Mappings will be performed every 24 hours (will be a changeable property).
+The Key to change this value using Spring properties is `toscana.docker.base-image-mapper.update-interval`
+
+The Mapping works as follows:
+1. The mapper checks if the given capability is empty (all optionals do not conatin a value). If thats the case, it returns the default image (`library/ubuntu:latest`, the latest Ubuntu LTS version (currently 16.04))
+2. The type attribute of the Capability gets checked. If the value is not set or is `linux` the mapping will continue and fail otherwise.
```
[the remark this answers is not in the document]
Comment (c-mueller) [reply docs/dev/plugins/kubernetes/dockerfile-mapping/base-image-mapper.md]: Yes this can still be done. I didnt want to miss this here just for completeness.
